x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/GrokPatternCreator.java

```diff
@@ -606,16 +617,22 @@ public String getTimeField() {
             return timeField;
         }
 
-        public boolean matchesAll(Collection<String> sampleMessages) {
-            return sampleMessages.stream().allMatch(grok::match);
+        public boolean matchesAll(Collection<String> sampleMessages, TimeoutChecker timeoutChecker) {
+            for (String sampleMessage : sampleMessages) {
+                if (grok.match(sampleMessage) == false) {
+                    return false;
+                }
+                timeoutChecker.check("full message Grok pattern matching");
```

Comment:
Can this be moved outside the loop? How many sample messages does the call expect?

Reply:
Grok pattern matching is one of the most expensive operations.  In Ingest pipeline a watchdog that's a little bit similar to the `TimeoutChecker` of this PR is installed to possibly interrupt every single cal to `match()`: https://github.com/elastic/elasticsearch/blob/ba3ceeaccf9b59ad16af60e279dbd6339e091a48/libs/grok/src/main/java/org/elasticsearch/grok/Grok.java#L226

With the structure finder's use of Grok that inner watchdog is the no-op watchdog, so we're not double checking for slow Grok patterns.  But it still seems reasonable to check once per Grok pattern match given that it's such an expensive operation.